src/devices/Hcsr04/Hcsr04.cs

```diff
@@ -42,13 +42,39 @@ public Hcsr04(int triggerPin, int echoPin, PinNumberingScheme pinNumberingScheme
             _controller.OpenPin(_trigger, PinMode.Output);
 
             _controller.Write(_trigger, PinValue.Low);
+
+            // Call Read once to make sure method is JITted
+            // Too long JITting is causing that initial echo pulse is frequently missed on the first run
+            // which would cause unnecessary retry
+            _controller.Read(_echo);
         }
 
         /// <summary>
         /// Gets the current distance in cm.
         /// </summary>
         private double GetDistance()
         {
+            // Retry at most 3 times.
+            // Try method will fail when context switch occurs in the wrong moment
+            // or something else (i.e. JIT, extra workload) causes extra delay
+            for (int i = 0; i < 3; i++)
+            {
+                if (TryGetDistance(out double result))
+                {
+                    return result;
+                }
+            }
+
+            throw new InvalidOperationException("Could not get reading from the sensor");
+        }
+
+        private bool TryGetDistance(out double result)
+        {
+            // Time when we give up on looping and declare that reading failed
+            // 100ms was chosen because max measurement time for this sensor is around 24ms for 400cm
+            // additionally we need to account 60ms max delay.
+            // Rounding this up to a 100 in case of a context switch.
+            long hangTicks = Environment.TickCount + 100;
```

Comment:
`Environment.TickCount` is a 32 bit value that overflows after 23 or so days. So this will fail when the value overflows, causing the delay to be 0. Netcore 3.0 or later has a `TickCount64` property as a possible workaround, but I think we can't use that yet.

Reply:
Good point, while this would retry it in that case and still work it's not a good solution. `Environment.TickCount - hangTicks > 0` seems to be working fine though since subtraction will "undo" overflowing